sync2/atxs.go

```diff
@@ -0,0 +1,338 @@
+}
+
+var _ multipeer.SyncKeyHandler = &ATXHandler{}
+
+func NewATXHandler(
+	logger *zap.Logger,
+	f Fetcher,
+	batchSize, maxAttempts, maxBatchRetries int,
+	failedBatchDelay time.Duration,
+	clock clockwork.Clock,
+) *ATXHandler {
+	if clock == nil {
+		clock = clockwork.NewRealClock()
+	}
+	return &ATXHandler{
+		f:                f,
+		logger:           logger,
+		clock:            clock,
+		batchSize:        batchSize,
+		maxAttempts:      maxAttempts,
+		maxBatchRetries:  maxBatchRetries,
+		failedBatchDelay: failedBatchDelay,
+	}
+}
+
+type commitState struct {
+	state         map[types.ATXID]int
+	total         int
+	numDownloaded int
+	items         []types.ATXID
+}
+
+func (h *ATXHandler) setupState(
+	peer p2p.Peer,
+	base rangesync.OrderedSet,
+	received rangesync.SeqResult,
+) (*commitState, error) {
+	state := make(map[types.ATXID]int)
+	for k := range received.Seq {
+		found, err := base.Has(k)
+		if err != nil {
+			return nil, fmt.Errorf("check if ATX exists: %w", err)
+		}
+		if found {
+			continue
+		}
+		id := types.BytesToATXID(k)
+		h.f.RegisterPeerHash(peer, id.Hash32())
+		state[id] = 0
+	}
+	if err := received.Error(); err != nil {
+		return nil, fmt.Errorf("get item: %w", err)
+	}
+	return &commitState{
+		state: state,
+		total: len(state),
+		items: make([]types.ATXID, 0, h.batchSize),
+	}, nil
+}
+
+func (h *ATXHandler) getAtxs(ctx context.Context, cs *commitState) (bool, error) {
+	cs.items = cs.items[:0] // reuse the slice to reduce allocations
+	for id := range cs.state {
+		cs.items = append(cs.items, id)
+		if len(cs.items) == h.batchSize {
+			break
+		}
+	}
+	someSucceeded := false
+	var mtx sync.Mutex
+	err := h.f.GetAtxs(ctx, cs.items, system.WithATXCallback(func(id types.ATXID, err error) {
+		mtx.Lock()
+		defer mtx.Unlock()
+		switch {
+		case err == nil:
+			cs.numDownloaded++
+			someSucceeded = true
+			delete(cs.state, id)
+		case errors.Is(err, pubsub.ErrValidationReject):
+			h.logger.Debug("failed to download ATX",
+				zap.String("atx", id.ShortString()), zap.Error(err))
+			delete(cs.state, id)
+		case cs.state[id] >= h.maxAttempts-1:
+			h.logger.Debug("failed to download ATX: max attempts reached",
+				zap.String("atx", id.ShortString()))
+			delete(cs.state, id)
+		default:
+			cs.state[id]++
+		}
+	}))
+	return someSucceeded, err
+}
+
+func (h *ATXHandler) Commit(
+	ctx context.Context,
+	peer p2p.Peer,
+	base rangesync.OrderedSet,
+	received rangesync.SeqResult,
+) error {
+	h.logger.Debug("begin atx commit")
+	defer h.logger.Debug("end atx commit")
+	cs, err := h.setupState(peer, base, received)
+	if err != nil {
+		return err
+	}
+	startTime := h.clock.Now()
+	batchAttemptsRemaining := h.maxBatchRetries
+	for len(cs.state) > 0 {
+		someSucceeded, err := h.getAtxs(ctx, cs)
+		switch {
+		case err == nil:
+		case errors.Is(err, context.Canceled):
+			return err
+		case !errors.Is(err, &fetch.BatchError{}):
+			h.logger.Debug("failed to download ATXs", zap.Error(err))
+		}
```

Comment:
Sorry, I had this wrong in my previous suggestion. `errors.Is` is used to check for value equality, not type equality. We probably want to do this here instead:

```suggestion
		someSucceeded, err := h.getAtxs(ctx, cs)
		batchErr := &fetch.BatchError{}
		switch {
		case err == nil:
		case errors.Is(err, context.Canceled):
			return err
		case !errors.As(err, batchErr):
			h.logger.Debug("failed to download ATXs", zap.Error(batchErr))
		}
```

The reason being that `errors.As` will check (and assign) `err` to `batchErr` if the type allows it. `errors.Is` does an equality check which by default will always fail unless `err == &fetch.BatchError{}`. If the type implements `interface{ Is(error) bool }` then `errors.Is` will use the `Is` method instead. `fetch.BatchError` has such a method that correctly checks if `errors.Is` would be `true` for any of the contained errors so that a caller can treat a batch error as if it was an error of a single item fetched.

Reply:
Fixed, but it had to be `case !errors.As(err, &batchErr):` b/c `BatchError` methods have pointer receiver